Add Upgrade system and fix typos

henriiquecampos opened the pull request:
**Please check if the PR fulfills these requirements:**

- [x] The commit message follows our guidelines.
- For bug fixes and features:
    - [x] You tested the changes.
    - [ ] You updated the docs or changelog.

**What kind of change does this PR introduce?**
Implements the Upgrade System, also fix some typos and change a small detail in the `Tower.gd`. Reason for that is that as the UpgradePanel isn't part of the Tower anymore, just showing the weapon's fire range should be enough when players buy a tower.


**Does this PR introduce a breaking change?**
No.


## New feature or change ##


**What is the current behavior?** 
Currently, players can't upgrade towers.


**What is the new behavior?**
With this PR, players can upgrade the RifleTower's fire range, also designers can create new upgrades.

henriiquecampos committed:
Add Upgrade system and fix typos

NathanLovato commented on source/UI/UITowerShop/UITowerShop.gd:
```diff
@@ -35,7 +35,7 @@ func _on_Tower_sold(price: int, _position: Vector2) -> void:
 func _on_Player_gold_changed(gold_amount: int) -> void:
 	for button in _buttons:
-		if button.tower_cost > player.gold:
+		if button.tower_cost > gold_amount:
 			button.disabled = true
 		else:
 			button.disabled = false
```
I always forget but these are good candidates for Boolean expressions. 

`button.disabled = button.tower_cost > gold_amount`

NathanLovato commented:
Looking good at a glance, I'll open it in Godot to check the code in detail and the menu issue you mentioned.

NathanLovato commented on source/Objects/Towers/Upgrades/Upgrade.gd:
```diff
@@ -0,0 +1,23 @@
+export(String, MULTILINE) var description := "Upgrade description"
+export var cost := 100
+export(float, EASE) var cost_curve := 1.0
+export var value := 100.0
```
Is it unused? If so, best remove it.

NathanLovato commented on source/UI/UIUpgradePanel/UIUpgradeShop.gd:
```diff
@@ -0,0 +1,62 @@
+	
+	var upgrades := tower.get_upgrades()
+	for upgrade in upgrades.get_children():
+		add_upgrade_option(upgrade)
```
Here the get_upgrades function might as well return the array of upgrade objects.

NathanLovato commented on source/UI/UIUpgradePanel/UIUpgradeShop.gd:
```diff
@@ -0,0 +1,62 @@
+
+	button.connect("pressed", self, "_on_Button_pressed", [button])
+	button.upgrade = upgrade
+	button.disabled = player.gold < upgrade.cost
```
You can directly bind the `upgrade` to the call, that way, you don't have to store any reference to the upgrade.

NathanLovato committed:
Refactor upgrades code

Simplify UIUpgradeButton and UIUpgradeShop
Rename Upgrade properties to be more descriptive

NathanLovato commented on source/Objects/Towers/Upgrades/Upgrade.gd:
```diff
@@ -0,0 +1,23 @@
+export var text := "Upgrade name"
+export(String, MULTILINE) var description := "Upgrade description"
+export var cost := 100
+export(float, EASE) var cost_curve := 1.0
```
You have to use the `ease` function to use such a value. You're using it as a multiplier, so I went ahead and removed the hint + renamed the property.

NathanLovato reviewed (approved):
I've made some requested changes. I'll let you check the comments to see what they're about. They simplify the UI code.

I found 4 bugs in the game, I didn't fix them because the changes may affect the lessons you wrote already. I'll open issues next.
